Assets/Tests/Runtime/ConnectionManagementTests.cs

```diff
@@ -0,0 +1,593 @@
+                Assert.AreEqual(expectedClientConnectionStateSequence, m_ClientConnectionStateSequences[i]);
+            }
+        }
+
+        [UnityTest]
+        public IEnumerator UserRequestedHostShutdownAfterClientsConnected_ClientsDisconnectedWithReason()
+        {
+            StartHost();
+            Assert.IsTrue(m_ServerNetworkManager.IsHost);
+
+            SetUniqueProfilesForEachClient();
+
+            yield return ConnectClients();
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                Assert.IsTrue(m_ClientNetworkManagers[i].IsConnectedClient);
+            }
+
+            m_ServerConnectionManager.RequestShutdown();
+
+            yield return new WaitWhile(() => m_ServerNetworkManager.IsListening);
+
+            Assert.IsFalse(m_ServerNetworkManager.IsHost);
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var clientId = i;
+                yield return new WaitWhile(() => m_ClientNetworkManagers[clientId].IsListening);
+                Assert.IsFalse(m_ClientNetworkManagers[clientId].IsConnectedClient);
+            }
+
+            var expectedServerConnectionStateSequence = new List<ConnectionState>();
+            expectedServerConnectionStateSequence.Add(m_ServerConnectionManager.m_StartingHost);
+            expectedServerConnectionStateSequence.Add(m_ServerConnectionManager.m_Hosting);
+            expectedServerConnectionStateSequence.Add(m_ServerConnectionManager.m_Offline);
+            Assert.AreEqual(expectedServerConnectionStateSequence, m_ServerConnectionStateSequence);
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var expectedClientConnectionStateSequence = new List<ConnectionState>();
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_ClientConnecting);
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_ClientConnected);
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_DisconnectingWithReason);
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_Offline);
+                Assert.AreEqual(expectedClientConnectionStateSequence, m_ClientConnectionStateSequences[i]);
+            }
+        }
+
+        [UnityTest]
+        public IEnumerator AttemptingToConnectWithSamePlayerId_ClientsDisconnectedWithReason()
+        {
+            StartHost();
+            Assert.IsTrue(m_ServerNetworkManager.IsHost);
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var profileManager = m_ClientScopes[i].Container.Resolve<ProfileManager>();
+                profileManager.Profile = $"Client";
+            }
+
+            yield return ConnectClients();
+
+            Assert.IsTrue(m_ClientNetworkManagers[0].IsConnectedClient);
+            for (var i = 1; i < NumberOfClients; i++)
```

Comment:
You might want to add some comments around this, from an external point of view, it's hard to read this test without knowing that implicitly this uses the fact that we can connect only once with the same profile.